### PR TITLE
fixed reconnection issue

### PR DESCRIPTION
**Pull request by robphilipp-tiggee**
Updated the SocketWriter to set closed socket writers and connections to nil so that the nil check is invoked to reconnect.

### CODE REVIEW

**Comment by deathowl** [comment]
:+1: @cmikk could you please review this?

File: SocketWriter.go [hunk 1]
```diff
@@ -181,14 +181,18 @@ func (sw *socketWriter) Close() error {
...
 		if err == nil {
 			return sw.c.Close()
 		}
 		sw.c.Close()
+		sw.c = nil
 		return err
 	}
 	if sw.c != nil {
-		return sw.c.Close()
+		var err = sw.c.Close()
```
Comment (vvfedorenko) [inline SocketWriter.go:193]: I believe `err` is visible in this scope and could be reused.

File: SocketWriter.go [hunk 2]
```diff
@@ -181,14 +181,18 @@ func (sw *socketWriter) Close() error {
 	var err error
 	if sw.w != nil {
 		err = sw.w.Close()
+		sw.w = nil
 		if err == nil {
 			return sw.c.Close()
```
Comment (vvfedorenko) [inline SocketWriter.go:186]: should we clean `sw.c` in this code-path too?

**Comment by cmikk** [comment]
Thanks all for your contributions and comments, and for your patience as I make time to review these.

Looking at this and @deathowl 's fix in #47 , it occurs to me that using `sw.w == nil` as a check for whether the writer has an open connection isn't ideal, as there's no other reason to `nil` out a writer or connection after closing it. In #48, I add an explicit flag to the socketWriter struct to indicate whether an open or reopen is needed.

In my tests, this appears to reconnect successfully. Can you, @deathowl, or @vvfedorenko confirm the fix in #48 works in your setup?

Thanks,
-Chris

**Comment by deathowl** [comment]
Yes , I verified #48 fixes the issue properly

**Comment by cmikk** [comment]
Fixed in #48

**Pull request closed without merging**